From bfa033b07f50a20229f33479c1058a3c7c22453f Mon Sep 17 00:00:00 2001
From: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Thu, 13 Apr 2023 14:42:06 -0300
Subject: [PATCH 3/5] pdwtags: Print DW_TAG_constant tags

First seen in Go compile units:

  $ readelf -wi main  | grep DW_AT_language -m1
      <33>   DW_AT_language    : 22	(Go)
  $ pdwtags main | grep Constants: -A20
  /* Constants: */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypeA = 1; /* size: 2 */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypeNS = 2; /* size: 2 */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypeCNAME = 5; /* size: 2 */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypeSOA = 6; /* size: 2 */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypePTR = 12; /* size: 2 */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypeMX = 15; /* size: 2 */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypeTXT = 16; /* size: 2 */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypeAAAA = 28; /* size: 2 */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypeSRV = 33; /* size: 2 */

  const vendor/golang.org/x/net/dns/dnsmessage.Type vendor/golang.org/x/net/dns/dnsmessage.TypeOPT = 41; /* size: 2 */
  $

Signed-off-by: Arnaldo Carvalho de Melo <acme@redhat.com>
---
 pdwtags.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/pdwtags.c b/pdwtags.c
index 8b1d6f1..5b62218 100644
--- a/pdwtags.c
+++ b/pdwtags.c
@@ -67,6 +67,11 @@ static int cu__emit_tags(struct cu *cu)
 		printf(" /* size: %zd */\n\n", tag__size(tag, cu));
 	}
 
+	puts("\n\n/* Constants: */\n");
+	cu__for_each_constant(cu, i, tag) {
+		tag__fprintf(tag, cu, NULL, stdout);
+		printf(" /* size: %zd */\n\n", tag__size(tag, cu));
+	}
 
 	return 0;
 }
-- 
2.39.1

